File: languages/python3/383_ransom-note.py
# See: https://leetcode.com/problems/ransom-note/solutions/3856466/python3-using-dict-as-a-hash-map/


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        cache = dict()  # create an empty dict to be used later
        # Create hash-map
        for letter in magazine:
            if letter in cache:
                cache[letter] += 1
            else:
                cache[letter] = 1

        # Do check
        for letter in ransomNote:
            if letter not in cache:
                return False
            elif cache[letter] == 0:
                return False
            else:
                cache[letter] -= 1

        return True


# Time complexity: O(n)
# Space complexity: O(n)

File: languages/java/383_ransom-note.java
// See: https://leetcode.com/explore/learn/card/the-leetcode-beginners-guide/692/challenge-problems/4427/

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> magazineLetters = new HashMap<>(); // HashMap to store characters as numbers

        for (int i = 0; i < magazine.length(); i++) {
            char m = magazine.charAt(i);

            int currentCount = magazineLetters.getOrDefault(m, 0);
            magazineLetters.put(m , currentCount + 1); // Update the frequency of the letter to the HashMap
        }

        for (int i = 0; i < ransomNote.length(); i++) {
            char r = ransomNote.charAt(i);

            int currentCount = magazineLetters.getOrDefault(r, 0);

            if (currentCount == 0) { // If the letter is not found in magazine, return immediately
                return false;
            }

            magazineLetters.put(r, currentCount - 1); // If it is found, update the frequency of the letter, like removing it from its pile
        }

        return true;

    }

}

// Time complexity: O(m)
// Space complexity: O(k)
